### PR TITLE
Luv 0.5.11: binding to libuv, the async I/O library

### PR DESCRIPTION
**Pull request by aantron**
From the [changelog](https://github.com/aantron/luv/releases/tag/0.5.11):

> Additions
>
> - Upgrade libuv to [1.43.0](https://github.com/libuv/libuv/releases/tag/v1.43.0) (aantron/luv#128).
> - Support building on MSVC (aantron/luv#126, Jonah Beckford).
>
> Bugs fixed
>
> - Invalid pointer-integer comparison warning on Windows (aantron/luv#127, Antonin Décimo).

Besides Luv 0.5.11, this PR adds package `luv_unix` to the opam repository for the first time. `luv_unix` is a [tiny adapter](https://github.com/aantron/luv/blob/2dc38710097a598beab942e0cb0ffe41d20ec94d/src/unix/luv_unix.mli) between Luv and the standard library's `Unix` (Luv is, otherwise, a completely standalone I/O library that could be used without `Unix`, in principle).

cc @talex5, @jonahbeckford

### CODE REVIEW

File: packages/luv_unix/luv_unix.0.5.0/opam [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+authors: "Anton Bachin <antonbachin@yahoo.com>"
+maintainer: "Anton Bachin <antonbachin@yahoo.com>"
+dev-repo: "git+https://github.com/aantron/luv.git"
+
+depends: [
+  "base-unix"
+  "dune" {>= "2.0.0"}
+  "luv"
+  "ocaml" {>= "4.02.0"}
```
Comment (kit-ty-kate) [inline packages/luv_unix/luv_unix.0.5.0/opam:18]:
```suggestion
  "ocaml" {>= "4.03.0"}
```
```
#=== ERROR while compiling luv_unix.0.5.0 =====================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.02.3 | pinned(https://github.com/aantron/luv/releases/download/0.5.11/luv-0.5.11.tar.gz)
# path                 ~/.opam/4.02/.opam-switch/build/luv_unix.0.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p luv_unix -j 31
# exit-code            1
# env-file             ~/.opam/log/luv_unix-15125-d2c37b.env
# output-file          ~/.opam/log/luv_unix-15125-d2c37b.out
### output ###
#       ocamlc src/unix/.luv_unix.objs/byte/luv_unix.{cmi,cmti} (exit 2)
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I src/unix/.luv_unix.objs/byte -I /home/opam/.opam/4.02/lib/bigarray-compat -I /home/opam/.opam/4.02/lib/bytes -I /home/opam/.opam/4.02/lib/ctypes -I /home/opam/.opam/4.02/lib/integers -I /home/opam/.opam/4.02/lib/luv -I /home/opam/.opam/4.02/lib/luv/c -I /home/opam/.opam/4.02/lib/luv/c_function_descriptions -I /home/opam/.opam/4.02/lib/luv/c_type_descriptions -I /home/opam/.opam/4.02/lib/ocaml/threads -I /home/opam/.opam/4.02/lib/result -no-alias-deps -o src/unix/.luv_unix.objs/byte/luv_unix.cmi -c -intf src/unix/luv_unix.mli)
# File "src/unix/luv_unix.mli", line 14, characters 39-58:
# Error: Unbound type constructor result
```

File: packages/luv_unix/luv_unix.0.5.0/opam [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  "result"
```
Comment (kit-ty-kate) [inline packages/luv_unix/luv_unix.0.5.0/opam:19]: If the previous comment is expected, maybe you don’t need `result` here.

File: packages/luv_unix/luv_unix.0.5.0/opam [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+bug-reports: "https://github.com/aantron/luv/issues"
+
+authors: "Anton Bachin <antonbachin@yahoo.com>"
+maintainer: "Anton Bachin <antonbachin@yahoo.com>"
+dev-repo: "git+https://github.com/aantron/luv.git"
+
+depends: [
+  "base-unix"
+  "dune" {>= "2.0.0"}
+  "luv"
```
Comment (kit-ty-kate) [inline packages/luv_unix/luv_unix.0.5.0/opam:17]:
```suggestion
  "luv" {= version}
```
```
#=== ERROR while compiling luv_unix.0.5.0 =====================================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.04.2 | pinned(https://github.com/aantron/luv/releases/download/0.5.11/luv-0.5.11.tar.gz)
# path                 ~/.opam/4.04/.opam-switch/build/luv_unix.0.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p luv_unix -j 31
# exit-code            1
# env-file             ~/.opam/log/luv_unix-15354-585b75.env
# output-file          ~/.opam/log/luv_unix-15354-585b75.out
### output ###
#       ocamlc src/unix/luv_unix.o (exit 2)
# (cd _build/default/src/unix && /home/opam/.opam/4.04/bin/ocamlc.opt -g -I /home/opam/.opam/4.04/lib/bytes -I /home/opam/.opam/4.04/lib/ctypes -I /home/opam/.opam/4.04/lib/integers -I /home/opam/.opam/4.04/lib/luv -I /home/opam/.opam/4.04/lib/luv/c -I /home/opam/.opam/4.04/lib/luv/c_function_descriptions -I /home/opam/.opam/4.04/lib/luv/c_type_descriptions -I /home/opam/.opam/4.04/lib/ocaml/threads -I /home/opam/.opam/4.04/lib/result -ccopt -std=gnu99 -ccopt -O2 -ccopt -fno-strict-aliasing -ccopt -fwrapv -ccopt -fcommon -ccopt -Wall -ccopt -D_FILE_OFFSET_BITS=64 -ccopt -D_REENTRANT -ccopt -fPIC -ccopt -g -o luv_unix.o luv_unix.c)
# luv_unix.c:9:10: fatal error: uv.h: No such file or directory
#     9 | #include <uv.h>
#       |          ^~~~~~
# compilation terminated.
#       ocamlc src/unix/.luv_unix.objs/byte/luv_unix.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.04/bin/ocamlc.opt -w -40 -g -bin-annot -I src/unix/.luv_unix.objs/byte -I /home/opam/.opam/4.04/lib/bytes -I /home/opam/.opam/4.04/lib/ctypes -I /home/opam/.opam/4.04/lib/integers -I /home/opam/.opam/4.04/lib/luv -I /home/opam/.opam/4.04/lib/luv/c -I /home/opam/.opam/4.04/lib/luv/c_function_descriptions -I /home/opam/.opam/4.04/lib/luv/c_type_descriptions -I /home/opam/.opam/4.04/lib/ocaml/threads -I /home/opam/.opam/4.04/lib/result -intf-suffix .ml -no-alias-deps -o src/unix/.luv_unix.objs/byte/luv_unix.cmo -c -impl src/unix/luv_unix.ml)
# File "src/unix/luv_unix.ml", line 22, characters 9-50:
# Error: Unbound value C.Functions.Os_fd.is_invalid_handle_value
#     ocamlopt src/unix/.luv_unix.objs/native/luv_unix.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.04/bin/ocamlopt.opt -w -40 -g -I src/unix/.luv_unix.objs/byte -I src/unix/.luv_unix.objs/native -I /home/opam/.opam/4.04/lib/bytes -I /home/opam/.opam/4.04/lib/ctypes -I /home/opam/.opam/4.04/lib/integers -I /home/opam/.opam/4.04/lib/luv -I /home/opam/.opam/4.04/lib/luv/c -I /home/opam/.opam/4.04/lib/luv/c_function_descriptions -I /home/opam/.opam/4.04/lib/luv/c_type_descriptions -I /home/opam/.opam/4.04/lib/ocaml/threads -I /home/opam/.opam/4.04/lib/result -intf-suffix .ml -no-alias-deps -o src/unix/.luv_unix.objs/native/luv_unix.cmx -c -impl src/unix/luv_unix.ml)
# File "src/unix/luv_unix.ml", line 22, characters 9-50:
# Error: Unbound value C.Functions.Os_fd.is_invalid_handle_value
```

**Comment by kit-ty-kate** [comment]
luv_unix is still failing with:
```
#=== ERROR while compiling luv_unix.0.5.0 =====================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.03.0 | pinned(https://github.com/aantron/luv/releases/download/0.5.11/luv-0.5.11.tar.gz)
# path                 ~/.opam/4.03/.opam-switch/build/luv_unix.0.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p luv_unix -j 31
# exit-code            1
# env-file             ~/.opam/log/luv_unix-7557-865888.env
# output-file          ~/.opam/log/luv_unix-7557-865888.out
### output ###
#          gcc src/unix/luv_unix.o (exit 1)
# (cd _build/default/src/unix && /usr/bin/gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fcommon -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fcommon -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC -g -I /home/opam/.opam/4.03/lib/ocaml -I /home/opam/.opam/4.03/lib/bigarray-compat -I /home/opam/.opam/4.03/lib/bytes -I /home/opam/.opam/4.03/lib/ctypes -I /home/opam/.opam/4.03/lib/integers -I /home/opam/.opam/4.03/lib/luv -I /home/opam/.opam/4.03/lib/luv/c -I /home/opam/.opam/4.03/lib/luv/c_function_descriptions -I /home/opam/.opam/4.03/lib/luv/c_type_descriptions -I /home/opam/.opam/4.03/lib/ocaml/threads -I /home/opam/.opam/4.03/lib/result -o luv_unix.o -c luv_unix.c)
# luv_unix.c:9:10: fatal error: uv.h: No such file or directory
#     9 | #include <uv.h>
#       |          ^~~~~~
# compilation terminated.
#       ocamlc src/unix/.luv_unix.objs/byte/luv_unix.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.03/bin/ocamlc.opt -w -40 -g -bin-annot -I src/unix/.luv_unix.objs/byte -I /home/opam/.opam/4.03/lib/bigarray-compat -I /home/opam/.opam/4.03/lib/bytes -I /home/opam/.opam/4.03/lib/ctypes -I /home/opam/.opam/4.03/lib/integers -I /home/opam/.opam/4.03/lib/luv -I /home/opam/.opam/4.03/lib/luv/c -I /home/opam/.opam/4.03/lib/luv/c_function_descriptions -I /home/opam/.opam/4.03/lib/luv/c_type_descriptions -I /home/opam/.opam/4.03/lib/ocaml/threads -I /home/opam/.opam/4.03/lib/result -intf-suffix .ml -no-alias-deps -o src/unix/.luv_unix.objs/byte/luv_unix.cmo -c -impl src/unix/luv_unix.ml)
# File "src/unix/luv_unix.ml", line 22, characters 9-50:
# Error: Unbound value C.Functions.Os_fd.is_invalid_handle_value
#     ocamlopt src/unix/.luv_unix.objs/native/luv_unix.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.03/bin/ocamlopt.opt -w -40 -g -I src/unix/.luv_unix.objs/byte -I src/unix/.luv_unix.objs/native -I /home/opam/.opam/4.03/lib/bigarray-compat -I /home/opam/.opam/4.03/lib/bytes -I /home/opam/.opam/4.03/lib/ctypes -I /home/opam/.opam/4.03/lib/integers -I /home/opam/.opam/4.03/lib/luv -I /home/opam/.opam/4.03/lib/luv/c -I /home/opam/.opam/4.03/lib/luv/c_function_descriptions -I /home/opam/.opam/4.03/lib/luv/c_type_descriptions -I /home/opam/.opam/4.03/lib/ocaml/threads -I /home/opam/.opam/4.03/lib/result -intf-suffix .ml -no-alias-deps -o src/unix/.luv_unix.objs/native/luv_unix.cmx -c -impl src/unix/luv_unix.ml)
# File "src/unix/luv_unix.ml", line 22, characters 9-50:
# Error: Unbound value C.Functions.Os_fd.is_invalid_handle_value
```

**Comment by aantron** [comment]
@kit-ty-kate I will re-tag the release and update this PR. It needs more invasive changes.

**Comment by talex5** [comment]
Does this even need to be a separate opam package? It doesn't require installing any extra opam dependencies compared to the main luv package, so maybe it could just be a `luv.unix` ocamlfind library, like we have with `lwt.unix`?

**Comment by kit-ty-kate** [comment]
Thanks! Could you return the fixes in the opam file upstream?

**Comment by aantron** [comment]
Thank you!

> Thanks! Could you return the fixes in the opam file upstream?

Of course. Done in https://github.com/aantron/luv/commit/0ed668e87ca55d65f4f84422b0dd3381701c71b0.

**Pull request merged**